src/hazelcore/ActionOutcome.re

```diff
@@ -1,4 +1,28 @@
-type t('success) =
-  | Succeeded('success)
-  | CursorEscaped(Side.t)
-  | Failed;
+module T = {
+  [@deriving sexp]
+  type t('success) =
+    | Succeeded('success)
+    | CursorEscaped(Side.t)
+    | Failed;
+
+  let map =
+    `Custom(
+      (x, f) =>
+        switch (x) {
+        | Succeeded(a) => Succeeded(f(a))
+        | CursorEscaped(s) => CursorEscaped(s)
```

Comment:
don't think we generally respond to CursorEscaped by passing it along -- each zipper case should decide how to respond to this situation in each sub-expression position carefully. suggests that maybe ActionOutcome shouldn't form a monad directly, but you could refactor it into an option(success) sort of thing

Reply:
okay I'll remove the monad instance on this for now, and refactor it to an option-like thing with the type variables work